### DISCUSSION TITLE
이벤트 루프의 동작과정을 실행컨텍스트와 연관지어 설명해주세요

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#73] [실행컨텍스트]
> 이벤트 루프의 동작과정을 실행컨텍스트와 연관지어 설명해주세요

**Comment by pipisebastian** [comment]
1. 자바스크립트은 하나의 콜스택을 가지는 싱글 스레드 언어지만, 실제로는 웹브라우저나 nodejs같은 런타임 환경에서 구동됩니다. 이러한 환경 덕분에 비동기 작업이 가능하게 되는데요.
    1. (대표적으로 setTimeout 같은 타이머, http 요청을 하는 ajax)
3. 즉, 자바스크립트의 싱글 스레드인 메인 스레드에서 브라우저의 멀티 스레드에 작업을 대신 요청하고, 이는 비동기적으로 수행됩니다. 여기서 작업이 끝나면, 메인 스레드가 콜백함수를 받아 결과를 실행하게 되는겁니다. 
4. 그리고 이러한 흐름을 관리하기 위해 **이벤트 루프**가 존재합니다. 웹 브라우저 상에서 동작과정을 간단하게 말씀드리겠습니다.
    1. 먼저 자바스크립트 엔진이 콜스택에 있는 함수들을 실행하다가, 비동기 함수인 setTimeOut가 실행되면, 브라우저의 web API를 통해 브라우저에게 작업을 요청하게 됩니다. 이때 브라우저의 멀티 스레드에서 작업이 수행됩니다.
        1. (web API가 모두 비동기적인 건 아님. console, DOM api는 동기적)
    2. 비동기 작업이 끝나면, 전달받았던 콜백함수를 콜백 큐에 넣습니다. 
    3. 이벤트 루프는 콜스택을 계속 확인하면서, 콜스택이 비는 순간에 콜백큐의 콜백함수를 콜스택에 넣고, 최종적으로 콜백함수를 실행하게 됩니다.

**Comment by publdaze** [comment]
> - 이벤트 루프의 동작 과정을 실행컨텍스트와 연관지어 설명해주세요
- 자바스크립트는 하나의 실행 컨텍스트를 가지는 싱글 스레드 기반 언어임
- 싱글 스레드 기반이므로 동기적으로 코드가 동작하는데, 비동기적으로 동작을 처리하고 싶을 때 이벤트 루프를 활용할 수 있음
- setTimeout 혹은 setInterver같은 비동기 작업의 콜백은 콜백 큐에 저장함
- 실행 컨텍스트는 함수 호출 시 생성되며, 콜 스택에 쌓이는데, 이벤트 루프는 이 콜 스택을 주시하다가 콜 스택이 비어 있을 때 콜백 큐에서 대기 중인 작업을 실행함
